psi4/src/psi4/scfgrad/jk_grad.cc

```diff
@@ -2332,6 +2350,61 @@ void DirectJKGrad::compute_gradient()
 
     if (!(Ca_ && Cb_ && Da_ && Db_ && Dt_))
         throw PSIEXCEPTION("Occupation/Density not set");
+    
+#ifdef USING_BrianQC
+    if (brianCookie != 0) {
+        brianBool computeCoulomb = (do_J_ ? BRIAN_TRUE : BRIAN_FALSE);
+        brianBool computeExchange = ((do_K_ || do_wK_) ? BRIAN_TRUE : BRIAN_FALSE);
+        bool betaFlag = (brianRestrictionType != BRIAN_RESTRICTION_TYPE_RHF);
+        
+        std::shared_ptr<Matrix> Jgrad, Kgrada, Kgradb;
+        if (computeCoulomb) {
+            Jgrad = std::make_shared<Matrix>("Coulomb Gradient", primary_->molecule()->natom(), 3);
+        }
+        if (computeExchange) {
+            Kgrada = std::make_shared<Matrix>("Exchange Gradient", primary_->molecule()->natom(), 3);
+            if (betaFlag) {
+                Kgradb = std::make_shared<Matrix>("Exchange Gradient beta", primary_->molecule()->natom(), 3);
+            }
+        }
+        
+        brianOPTBuildGradientRepulsionDeriv(&brianCookie,
+            &computeCoulomb,
+            &computeExchange,
+            Da_->get_pointer(),
+            (betaFlag ? Db_->get_pointer() : nullptr),
+            (computeCoulomb ? Jgrad->get_pointer() : nullptr),
+            (computeExchange ? Kgrada->get_pointer() : nullptr),
+            ((computeExchange && betaFlag) ? Kgradb->get_pointer() : nullptr)
+        );
+        
+        if (computeExchange) {
+            if (betaFlag) {
+                Kgrada->add(Kgradb);
+            } else {
+                Kgrada->scale(2.0);
+            }
+        }
+        
+        gradients_.clear();
+        
+        if (do_J_) {
+            gradients_["Coulomb"] = Jgrad;
+        }
+        
+        if (do_K_) {
+            gradients_["Exchange"] = Kgrada;
+            
+            if (do_wK_) {
+                gradients_["Exchange,LR"] = std::make_shared<Matrix>("Exchange,LR Gradient", primary_->molecule()->natom(), 3);
```

Comment:
Yes, it's getting initialized to zero. If there is both a "normal" and a range-separated exchange (do_K_ and do_wK_ both true), then BrianQC only computes their sum. (Also, it factors in the DFT functional's exchange weights internally, so Psi4's scaling later is disabled when BrianQC is enabled.) So, I wanted to ensure two things: that all matrices that Psi4 would otherwise initialize are still initialized (since I wasn't really sure if any later code depends on them existing), and that the sum of K and wK are always correct. So that's why in that branch, we put BrianQC's exchange (the sum of K and wK) into K, and we initialize wK, but leave it zero. It would be more accurate if BrianQC computed the two contributions separately, but the current API doesn't support that.

Reply:
got it. thanks for the explanation.